geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+
+        // create ring by collecting the points if something was found
+        let ring = points
+            .drain(pos..)
+            .chain(std::iter::once(p))
+            .cloned()
+            .collect::<Vec<_>>();
+        Some(ring)
+    }
+
+    // find rings
+    let rings = {
+        let (points, mut rings) =
+            poly.exterior()
+                .into_iter()
+                .fold((vec![], vec![]), |(mut points, mut rings), coord| {
+                    if let Some(ring) = detect_if_rings_closed_with_point(&mut points, coord) {
+                        rings.push(ring);
+                    }
+                    points.push(coord);
+                    (points, rings)
+                });
+
+        // add leftover coords as last ring
+        let last_ring = points.into_iter().cloned().collect::<Vec<_>>();
+        rings.push(last_ring);
+
+        rings
+    };
+
+    // convert to polygons for containment checks
+    let mut rings = rings
+        .into_iter()
+        // filter out degenerate polygons which may be produced from the code above
+        .filter(|cs| cs.len() >= 3)
+        .map(|cs| Polygon::new(LineString::new(cs), vec![]))
+        .collect::<Vec<_>>();
+
+    fn find_outmost_ring<F: GeoFloat>(rings: &[Polygon<F>]) -> Option<usize> {
+        let enumerated_rings = || rings.iter().enumerate();
+        enumerated_rings()
+            .find(|(i, ring)| {
+                enumerated_rings()
+                    .filter(|(j, _)| i != j)
+                    .all(|(_, other)| ring.contains(other))
+            })
+            .map(|(i, _)| i)
+    }
+
+    // if exterior ring exists that contains all other rings, recreate the poly with:
+    //
+    // - exterior ring as exterior
+    // - other rings are counted to interiors
+    // - previously existing interiors are preserved
+    if let Some(outer_index) = find_outmost_ring(&rings) {
+        let exterior = rings.remove(outer_index).exterior().clone();
+        let interiors = poly
+            .interiors()
+            .iter()
+            .cloned()
+            .chain(rings.into_iter().map(|p| p.exterior().clone()))
+            .collect::<Vec<_>>();
+        poly = Polygon::new(exterior, interiors);
+    }
+    poly
+}
+
+/// Inputs to this function is a unordered set of lines that must form a valid multipolygon
+fn stitch_multipolygon_from_lines<F: GeoFloat>(
+    lines: Vec<Line<F>>,
+) -> PolygonStitchingResult<MultiPolygon<F>> {
+    let rings = stitch_rings_from_lines(lines)?;
+
+    fn find_parent_idxs<F: GeoFloat>(
+        ring_idx: usize,
+        ring: &LineString<F>,
+        all_rings: &[LineString<F>],
+    ) -> Vec<usize> {
+        all_rings
+            .iter()
+            .enumerate()
+            .filter(|(other_idx, _)| ring_idx != *other_idx)
+            .filter_map(|(idx, maybe_parent)| {
+                Polygon::new(maybe_parent.clone(), vec![])
+                    .contains(ring)
```

Comment:
Uhh um, yeah I actually haven't thought about this one yet 🤔 Nice catch, I need to do some testing.

Reply:
Is this code strictly about trying to handle a "banana polygon". Arguably that's not ["valid" input](https://postgis.net/workshops/postgis-intro/validity.html), so I personally don't feel like we need to worry much about the output in that case — garbage in garbage out.

Or is there more to it than that?

Reply:
I finally had the time and the drive to review this question. I think the original question can be answered as follows:

If the inner ring "touches" the parent, the two linestrings that form the inner and outer ring will be stitched together. So this will result in a banana polygon. I'm not really sure whether or not this is desired by the user. Here's an visual example:

![image](https://github.com/georust/geo/assets/26892280/2e2147ee-43cf-49aa-8eec-89a9202b933c)

Currently if it were triangulated and we would stitch it together, we would end up with one multipolygon with one polygon with no hole. I'm not sure if the user would expect the inner thing to be a hole if it touches the outer ring

Reply:
@michaelkirk I believe this is one of two outstanding questions (along with https://github.com/georust/geo/pull/1087#discussion_r1476508440). WDYT? At this point, I'm inclined to document this as an edge case if need be in the interests of getting this trait merged so we can get #1089 merged.

Reply:
Thanks for the prod @urschrei. Im OK with documenting the expected output and moving on.

Reply:
I documented the case somewhat on the trait now :+1: